text/0162-refactoring-support.md

```diff
@@ -0,0 +1,665 @@
+                         │                  │ ┌─────────┐       │              
+                         └───────────────────>│ bucket  │       │              
+                                    after   │ └─────────┘       │              
+                                            └───────────────────┘
+
+In the CloudFormation templates, this is represented by an exported output in
+`Producer1`, which is referenced by an `Fn::ImportValue` in `Consumer`. What
+ultimately needs to happen is that the exported output in `Producer1` must be
+removed, and a new one must be added in `Producer2`. The `Fn::ImportValue` in
+`Consumer` must be updated to point to the new output.
+
+This is manifestation of the deadly embrace problem. We can't do this in a
+single step because it would create a dangling reference (and therefore
+CloudFormation would reject it). The solution is also similar to what we suggest
+for solving the deadly embrace, but is done automatically here:
+
+1. Update the value of the output in `Producer1` to the hard-coded physical ID
+   of the bucket.
+2. Call the refactor API to move the bucket from `Producer1` to `Producer2`,
+   preserving the old output in `Producer1` while creating a new output in
+   `Producer2`, that references the bucket.
+3. Update the `Fn::ImportValue` in `Consumer` to point to the new output in
+   `Producer2`.
+4. Remove the output in `Producer1`.
+
+### E. Similarity between resources
+
+There are cases in which developers may wish to move a resource, and, at the
+same time, modify some of its properties. Indeed, this may not even be a choice.
+Some CDK constructs, for example, add tags to the underlying resource that are
+based on the construct path. Intuitively, developers would expect that small
+changes like this would be taken into account by the matching algorithm.
+
+The current proposal, however, does not solve this problem, because it uses a
+digest function to compare resources. As such, given two resources, they are
+considered either equivalent or not, with no middle ground.
+
+So we need a more flexible way of comparing resources. Instead of requiring them
+to be exactly the same, we need to know whether two resources are "similar
+enough". To achieve this, we need a few things:
+
+- **A distance function**. To generalize the notion of strict equality to a more
+  flexible notion of similarity, we need to come up with a distance function,
+  that is, a function that takes two resources as input, and produces a number
+  as output. The higher the number, the more different the resources are.
+- **A threshold for the distance**. This is a cut-off value that we can use to
+  decide whether two resources are similar enough to be candidates for
+  refactoring. It will depend on the distance function we choose and also on
+  experimentation.
+- **Graph isomorphism**. Since we will not have a digest anymore to quickly find
```

Comment:
All very true, but I think we can probably wing it with a little user feedback and be *mostly* fine.

Take objects of the same type, order them by distance, ask the user to pick the mapping. Bob's your uncle 😎